[TensorExpr] Fix handling of 0-dim tensors.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59328 [TensorExpr] Fix printing of Bool dtype.
* #59280 [TensorExpr] Improve debug messages.
* **#59279 [TensorExpr] Fix handling of 0-dim tensors.**

There were some issues with how we handle 0-dim cases in lowerings and
also in how we generate reductions in that special case. This PR fixes
those issues and reenables a bunch of tests.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

Commit:
[TensorExpr] Enable fusion for 0-dim tensors.

NNC has actually supported 0-dim tensors out of the box despite out
impression that it doesn't. This PR fixes some issues in tests that
might have been perceived as a lack of 0-dim tensors support and enables
fusion of such tensors.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28819780).

Commit:
Update on "[TensorExpr] Enable fusion for 0-dim tensors."


NNC has actually supported 0-dim tensors out of the box despite out
impression that it doesn't. This PR fixes some issues in tests that
might have been perceived as a lack of 0-dim tensors support and enables
fusion of such tensors.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fix handling of 0-dim tensors."


There were some issues with how we handle 0-dim cases in lowerings and
also in how we generate reductions in that special case. This PR fixes
those issues and reenables a bunch of tests.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28819780).

Commit:
Update on "[TensorExpr] Fix handling of 0-dim tensors."


There were some issues with how we handle 0-dim cases in lowerings and
also in how we generate reductions in that special case. This PR fixes
those issues and reenables a bunch of tests.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28819780).

navahgar reviewed (commented):
Mostly LGTM. Couple of comments below.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2477,9 +2486,6 @@ Tensor* tensorexpr::computeOperandValue(
     }
     case aten::t: {
       auto shape = valueShape(inputs[0]);
-      if (shape.size() == 1) {
```
This is now handled in `aten::transpose` which we use for `aten::t`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2488,6 +2494,17 @@ Tensor* tensorexpr::computeOperandValue(
+            [&](std::vector<VarHandle> axes) {
+              TORCH_INTERNAL_ASSERT(axes.size() <= 1);
+              return A.load(axes);
+            });
```
On the second thought on this - I feel that such a method won't buy us much since we'd need to pass lots of things into it (A, outputShape, name_hint) and it will only be used in this one place. Would you be ok if we keep the code as is? The other two "trivial" computes are also different - they are accessing the 0th element of a 0-dim input tensor, this one works for 1-d and 0-d.

Commit:
rebase on "[TensorExpr] Fix handling of 0-dim tensors."


There were some issues with how we handle 0-dim cases in lowerings and
also in how we generate reductions in that special case. This PR fixes
those issues and reenables a bunch of tests.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28819780).

Commit:
Update on "[TensorExpr] Fix handling of 0-dim tensors."


There were some issues with how we handle 0-dim cases in lowerings and
also in how we generate reductions in that special case. This PR fixes
those issues and reenables a bunch of tests.

Differential Revision: [D28819780](https://our.internmc.facebook.com/intern/diff/D28819780)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28819780).
